receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,327 @@
+
+	// Set the proper target by creating one target inside a single target group (this is how Prometheus wants its scrape config)
+	scrapeConfig.ServiceDiscoveryConfig = sdconfig.ServiceDiscoveryConfig{
+		StaticConfigs: []*targetgroup.Group{
+			{
+				Targets: []model.LabelSet{
+					{model.AddressLabel: model.LabelValue(fmt.Sprintf("localhost:%v", cfg.Port))},
+				},
+			},
+		},
+	}
+
+	receiverSettings := &configmodels.ReceiverSettings{}
+	receiverSettings.SetType(typeStr)
+	receiverSettings.SetName(cfg.Name())
+
+	return &prometheusreceiver.Config{
+		ReceiverSettings: *receiverSettings,
+		PrometheusConfig: &config.Config{
+			ScrapeConfigs: []*config.ScrapeConfig{scrapeConfig},
+		},
+	}
+}
+
+// getSubprocessConfig returns the subprocess config
+func getSubprocessConfig(cfg *Config) *subprocessmanager.SubprocessConfig {
+	subprocessConfig := &subprocessmanager.SubprocessConfig{}
+
+	subprocessConfig.Command = cfg.SubprocessConfig.Command
+	subprocessConfig.Env = cfg.SubprocessConfig.Env
+
+	return subprocessConfig
+}
+
+// extractName will return the receiver's given custom name (prometheus_exec/custom_name)
+func extractName(cfg *Config) string {
+	splitName := strings.SplitN(cfg.Name(), "/", 2)
+	if len(splitName) > 1 && splitName[1] != "" {
+		return splitName[1]
+	}
+	// fall back to the first part of the string, prometheus_exec
+	return splitName[0]
+}
+
+// manageProcess will put the process in an infinite starting loop which goes through the following steps
+// If the port is not defined by the user, one is generated and a new metrics receiver is built with the new port
+// All instances of {{port}} are replaced with the actual port (either defined by the user or generated)
+// Start the receiver and then the subprocess and get its runtime, then shutdown receiver since it crashed
+// Finally, the wait time before the subprocess is restarted is computed and this goroutine sleeps for that amount of time, before restarting the loop from the start
+func (per *prometheusExecReceiver) manageProcess(ctx context.Context, host component.Host) {
+	var (
+		newPort int = per.originalPort
+
+		elapsed    time.Duration
+		crashCount int
+	)
+
+	for {
+
+		// Generate a port if none was specified
+		if per.originalPort == 0 {
+			var err error
+			newPort, err = per.assignNewRandomPort()
+			if err != nil {
+				per.logger.Error("assignNewRandomPort() error - killing this single process/receiver", zap.String("error", err.Error()))
+				return
+			}
+		}
+		per.subprocessConfig = per.fillPortPlaceholders(newPort)
+
+		// Create and start the underlying Prometheus receiver
+		factory := &prometheusreceiver.Factory{}
+		receiver, ok := factory.CreateMetricsReceiver(ctx, per.logger, per.promReceiverConfig, per.consumer)
+		if ok != nil {
+			per.logger.Error("unable to create Prometheus receiver - killing this single process/receiver", zap.String("error", ok.Error()))
+			return
+		}
+
+		per.prometheusReceiver = receiver
```

Comment:
Does this need to be stored against the struct actually? We only use it in this loop. We do need to make sure it gets shutdown properly, but that can be handled inside of `case shutdown := <-per.s:`